qlight/client_cache.go

```diff
@@ -0,0 +1,78 @@
+package qlight
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/private"
+	"github.com/ethereum/go-ethereum/private/cache"
+	"github.com/ethereum/go-ethereum/private/engine/qlightptm"
+	gocache "github.com/patrickmn/go-cache"
+)
+
+type clientCache struct {
+	cacheWithEmpty    CacheWithEmpty
+	privateBlockCache *gocache.Cache
+	db                ethdb.Database
+}
+
+func NewClientCache(db ethdb.Database) (PrivateClientCache, error) {
+	cachingTXManager, ok := private.P.(*qlightptm.CachingProxyTxManager)
+	if !ok {
+		return nil, fmt.Errorf("unable to initialize cacheWithEmpty")
+	}
+	return NewClientCacheWithEmpty(db, cachingTXManager, gocache.New(cache.DefaultExpiration, cache.CleanupInterval))
+}
+
+func NewClientCacheWithEmpty(db ethdb.Database, cacheWithEmpty CacheWithEmpty, gocache *gocache.Cache) (PrivateClientCache, error) {
+	return &clientCache{
+		cacheWithEmpty:    cacheWithEmpty,
+		privateBlockCache: gocache,
+		db:                db,
+	}, nil
+}
+
+func (c *clientCache) AddPrivateBlock(blockPrivateData BlockPrivateData) error {
+	for _, pvtTx := range blockPrivateData.PrivateTransactions {
+		if err := c.cacheWithEmpty.Cache(pvtTx.ToCachable()); err != nil {
+			return err
+		}
+	}
+	if !common.EmptyHash(blockPrivateData.PrivateStateRoot) {
+		return c.privateBlockCache.Add(blockPrivateData.BlockHash.ToBase64(), blockPrivateData.PrivateStateRoot.ToBase64(), gocache.DefaultExpiration)
+	}
+	return nil
+}
+
+func (c *clientCache) CheckAndAddEmptyEntry(hash common.EncryptedPayloadHash) {
+	c.cacheWithEmpty.CheckAndAddEmptyToCache(hash)
+}
+
+func (c *clientCache) ValidatePrivateStateRoot(blockHash common.Hash, publicStateRoot common.Hash) error {
+	dbPrivateStateRoot := rawdb.GetPrivateStateRoot(c.db, publicStateRoot)
+
+	cachePrivateStateRootStr, found := c.privateBlockCache.Get(blockHash.ToBase64())
+	if !found {
+		// this means that we don't have private data for this block or that the server does not have the corresponding
+		// private state root (which can happen when caching is enabled on the server side)
+		return nil
+	}
+	cachePrivateStateRootB64, ok := cachePrivateStateRootStr.(string)
+	if !ok {
+		return fmt.Errorf("Invalid private block cache item")
+	}
+	cachePrivateStateRoot, err := common.Base64ToHash(cachePrivateStateRootB64)
+	if err != nil {
+		return fmt.Errorf("Invalid encoding for private state root: %s", cachePrivateStateRootB64)
+	}
+	if !bytes.Equal(cachePrivateStateRoot.Bytes(), dbPrivateStateRoot.Bytes()) {
+		log.Error("QLight - Private state root hash check failure for block", "hash", blockHash)
```

Comment:
I've looked for the places where the error is handled and I could not find any `log.Error` calls. I've seen it handled in the Downloader (during peer sync) but it is only a `log.Debug` there (the assumption being that some peer is trying to push some shoddy data - in which case you disconnect from that peer and try your luck with others)...

Even though it may appear more than once in the logs I think it is important for now to have it as an error as part of the qlight code (in most cases people will only enable info logs). We can remove it later if people complain about it.